src/v/cluster/rm_stm.cc

```diff
@@ -423,11 +423,45 @@ ss::future<tx_errc> rm_stm::commit_tx(
     co_return tx_errc::none;
 }
 
+rm_stm::abort_origin rm_stm::get_abort_origin(
+  const model::producer_identity& pid, const model::tx_seq& tx_seq) const {
+    auto expected_it = _mem_state.expected.find(pid);
+    if (expected_it != _mem_state.expected.end()) {
+        if (tx_seq < expected_it->second) {
+            return abort_origin::past;
+        } else if (expected_it->second < tx_seq) {
```

Comment:
i think this is great; though at first i had to read twice. 

this is a nit, so no need to fix, but  seems easier for the reader to keep condition on the same side

```
if (tx_seq < ...) { past; }
if (tx_seq > ...) { future; }
// else we are at the right time :)
```

nit2: no need for `else if` since you are returning.